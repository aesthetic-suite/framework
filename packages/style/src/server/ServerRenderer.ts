import Renderer from '../Renderer';
import GlobalStyleSheet from '../GlobalStyleSheet';
import ConditionsStyleSheet from '../ConditionsStyleSheet';
import StandardStyleSheet from '../StandardStyleSheet';
import TransientStyleRule from './TransientStyleRule';

export default class ServerRenderer extends Renderer {
  protected globalStyleSheet = new GlobalStyleSheet(new TransientStyleRule(CSSRule.STYLE_RULE));

  protected conditionsStyleSheet = new ConditionsStyleSheet(
    new TransientStyleRule(CSSRule.STYLE_RULE),
  );

  protected standardStyleSheet = new StandardStyleSheet(new TransientStyleRule(CSSRule.STYLE_RULE));
}
