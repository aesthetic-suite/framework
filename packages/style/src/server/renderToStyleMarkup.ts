import ServerRenderer from './ServerRenderer';
import { StyleRule, SheetType } from '../types';

function createElement(type: SheetType, rule: StyleRule): string {
  let css = '';

  for (let i = 0; i < rule.cssRules.length; i += 1) {
    css += rule.cssRules[i].cssText;
  }

  return `<style id="aesthetic-${type}" type="text/css" media="screen" data-aesthetic-hydrate="true">${css}</style>`;
}

export default function renderToStyleMarkup(renderer: ServerRenderer): string {
  return (
    createElement('global', renderer.getRootRule('global')) +
    createElement('standard', renderer.getRootRule('standard')) +
    createElement('conditions', renderer.getRootRule('conditions'))
  );
}
