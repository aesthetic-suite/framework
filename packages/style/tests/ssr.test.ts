import ServerRenderer from '../src/server/ServerRenderer';
import renderToStyleMarkup from '../src/server/renderToStyleMarkup';

describe('SSR', () => {
  it('writes to a temporary style sheet implementation and generates accurate markup', () => {
    const renderer = new ServerRenderer();

    renderer.applyRootVariables({
      fontSize: '16px',
      bgColor: '#fff',
      fbColor: 'black',
    });

    renderer.renderRule({
      margin: 0,
      padding: '6px 12px',
      border: '1px solid #2e6da4',
      borderRadius: 4,
      display: 'inline-block',
      cursor: 'pointer',
      fontFamily: 'Roboto',
      fontWeight: 'normal',
      lineHeight: 'normal',
      whiteSpace: 'nowrap',
      textDecoration: 'none',
      textAlign: 'left',
      backgroundColor: '#337ab7',
      verticalAlign: 'middle',
      color: 'rgba(0, 0, 0, 0)',
      animationName: 'fade',
      animationDuration: '.3s',
    });

    renderer.renderFontFace({
      fontFamily: '"Open Sans"',
      fontStyle: 'normal',
      fontWeight: 800,
      src: 'url("fonts/OpenSans-Bold.woff2")',
    });

    renderer.renderRule({
      margin: 0,
      '@media (width: 500px)': {
        margin: 10,
        ':hover': {
          color: 'red',
        },
        '@media (width: 350px)': {
          '@supports (color: blue)': {
            color: 'blue',
          },
        },
      },
    });

    renderer.renderKeyframes({
      from: {
        transform: 'translateX(0%)',
      },
      to: {
        transform: 'translateX(100%)',
      },
    });

    renderer.renderImport('url("test.css")');

    expect(renderToStyleMarkup(renderer)).toMatchSnapshot();
  });
});
