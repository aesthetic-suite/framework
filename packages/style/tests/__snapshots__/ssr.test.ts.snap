// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR writes to a temporary style sheet implementation and generates accurate markup 1`] = `"<style id=\\"aesthetic-global\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-hydrate=\\"true\\">@font-face { font-family:\\"Open Sans\\"; font-style:normal; font-weight:800; src:url(\\"fonts/OpenSans-Bold.woff2\\"); }@keyframes kf103rcyx { from { transform:translateX(0%); } to { transform:translateX(100%); }  }</style><style id=\\"aesthetic-standard\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-hydrate=\\"true\\">.a { margin:0 }.b { padding:6px 12px }.c { border:1px solid #2e6da4 }.d { border-radius:4px }.e { display:inline-block }.f { cursor:pointer }.g { font-family:Roboto }.h { font-weight:normal }.i { line-height:normal }.j { white-space:nowrap }.k { text-decoration:none }.l { text-align:left }.m { background-color:#337ab7 }.n { vertical-align:middle }.o { color:rgba(0, 0, 0, 0) }.p { animation-name:fade }.q { animation-duration:.3s }</style><style id=\\"aesthetic-conditions\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-hydrate=\\"true\\">@media (width: 500px) { @media (width: 350px) { @supports (color: blue) { .t { color:blue } } } }@media (width: 500px) { .s:hover { color:red } }@media (width: 500px) { .r { margin:10px } }</style>"`;
